src/stdlib_ascii.fypp

```diff
@@ -312,4 +326,51 @@ pure function reverse(string) result(reverse_string)
 
     end function reverse
 
+    #:for kind in INT_KINDS
+    !> Represent an integer of kind ${kind}$ as character sequence
+    pure function to_string_integer_${kind}$(val) result(string)
+        integer, parameter :: ik = ${kind}$
+        integer(ik), intent(in) :: val
+        character(len=:), allocatable :: string
+        integer, parameter :: buffer_len = range(val)+2
+        character(len=buffer_len) :: buffer
+        integer :: pos
+        integer(ik) :: n
+        character(len=1), parameter :: numbers(0:9) = &
+            ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
+
+        if (val == 0_ik) then
+            string = numbers(0)
+            return
+        end if
+
+        n = abs(val)
+        buffer = ""
+
+        pos = buffer_len + 1
+        do while (n > 0_ik)
+            pos = pos - 1
+            buffer(pos:pos) = numbers(mod(n, 10_ik))
+            n = n/10_ik
+        end do
+        if (val < 0_ik) then
+            pos = pos - 1
+            buffer(pos:pos) = '-'
+        end if
+
+        string = buffer(pos:)
+    end function to_string_integer_${kind}$
```

Comment:
Is there a reason not to make this a bit simpler and just use internal file IO? (like [this](https://gitlab.com/everythingfunctional/strff/-/blob/main/src/strff.f90#L477))

Reply:
There was a brief discussion in https://github.com/fortran-lang/stdlib/issues/69#issuecomment-570563373 about internal IO vs alternative implementations and I got the impression that we want to avoid internal IO in such conversion routines.

Reply:
Skimming that thread I didn't see anything saying that. Maybe I just missed it.

I think at one point I heard rumblings about internal file IO being really slow, but I don't think I've ever seen anybody actually demonstrate it.

Reply:
> I think at one point I heard rumblings about internal file IO being really slow, but I don't think I've ever seen anybody actually demonstrate it.

My proposed [improvement to the Julia benchmarks](https://fortran-lang.discourse.group/t/improving-fortran-results-in-the-julia-micro-benchmarks/198) found a significant improvement in `itoa` type functionality by avoiding internal write with gfortran. After skimming the gfortran runtime library I eventually came to the conclusion that internal write just has __a lot__ of overhead because it must handle so many different eventualities.

Reply:
I did a quick benchmark at https://github.com/awvwgk/test-internal-io

The manual writing needs 3.9s while the internal IO needs 8.1s for the same workload (5 million integer to string conversions).

Reply:
Another reason, quoting from Jones & Crabtree (1988), Fortran Tools for VAX/VMS and MS-DOS, John Wiley & Sons: 

> The reason for not using internal files is mainly to avoid using Fortran I/O if possible on microcomputers with limited memory, with the hope that host compiler won't include Fortran I/O in the executable image; it is also kind of nice to be able to do it on your own. In our tests of IBM-PC program  that do not use Fortran I/O in the primitives, replacing versions of `itoa` and `atoi` using internal files with the ones shown here reduced the size of executable  programs by 21K bytes.

Reply:
Given that stdlib is both a specification and a *reference implementation* I think it is also fitting to demonstrate Fortran is capable of doing it's own string conversions efficiently (hopefully this turns out to be valid for floating point numbers too :roll_eyes:).

Reply:
I stand corrected. It looks good to me then.